### DISCUSSION TITLE
\*WORDLIST\*

### DISCUSSION THREAD

**Discussion by ghost** [tabemann/zeptoforth#154] [Q&A]
zeptoforth is a great piece of work!
By strolling through wordlist at I have fond about 132 items named \*WORDLIST* (the asterisks belong to this name).
What is the meaning of this \*WORDLIST\*s? Are they pointer to real wordlist (SPI, etc)? If so: How do I get the names of the wordlist and how do I get the names of the included items (words).
BTW: What are the differences between 'module', 'wordlist' and 'dict'
Thanks in advance!
Martin

**Comment by tabemann** [comment]
On Mon, Nov 4, 2024 at 3:20 AM MB ***@***.***> wrote:

> zeptoforth is a great piece of work!
> By strolling through wordlist at I have fond about 133 items named
> *WORDLIST* (the asterisks belong to this name).
> What is the meaning of this *WORDLIST*s? Are they pointer to real wordlist
> (SPI, etc)? If so: How do I get the names of the wordlist and how do I get
> the names of the included items (words).
>

All *WORDLIST*s are is constants that contain the index of the most
recently defined wordlist at that point, so that the next call to WORDLIST
can get the next wordlist. Normally WORDS filters these out (or at least
should) so it is interesting that you are seeing these constants -- are you
manually walking the list of words in the flash dictionary?

Note that there is no way to reverse-look up a wordlist/module's name,
because the words containing the id's of wordlists/modules are just normal
constants, and wordlist/module id's are just normal numeric constants
starting at 0 counting up for wordlists/modules in flash, and 32768
counting up for wordlists/modules in RAM. Also, it is trivial to make new
names for wordlists/modules; it can just be done with CONSTANT .


> BTW: What are the differences between 'module', 'wordlist' and 'dict'
> Thanks in advance!
> Martin
>

A 'module' and a 'wordlist' are effectively the same thing -- 'wordlist' is
just traditional Forth terminology, whereas 'module' refers to how they are
used within zeptoforth. As for a 'dict(ionary)', each task has an area of
RAM known as its RAM dictionary (with compilation to RAM typically going
into the RAM dictionary of the main task) while there is a flash dictionary
shared between all tasks which persists across reboots.

Travis

**Reply by ghost** [reply]
Thanks a lot! That made it clearer to me.
Yes - I'm going manually thru the the list of words in the flash and in the ram.
In nearly all forthes I use I write mwords and voc-mwords (close to your lookup). They get a pattern and list all words that matches this pattern. I think it is very usefull. It was copied under different names by other forthes ;-)
(see: https://theforth.net/package/mwords)

**Comment by tabemann** [comment]
On Mon, Nov 4, 2024 at 9:27 AM MB ***@***.***> wrote:

> Thanks a lot! That made it clearer to me.
> Yes - I'm going manually thru the the list of words in the flash and in
> the ram.
> In nearly all forthes I use I write mwords and voc-mwords (close to your
> lookup). They get a pattern and list all words that matches this pattern. I
> think it is very usefull. It was copied under different names by other
> forthes ;-)
> (see: https://theforth.net/package/mwords)
>
Note that words that start and end in * other than * itself are normally
supposed to be hidden from the user in word listings.

Also note that there is a convenient way to list words in a
wordlist/module, specifically WORDS-IN ( wid -- ) or, if you want a
screen-by-screen pager, MORE-WORDS-IN ( wid -- ). There is also a word for
listing all words in the current wordlist/module order with a
screen-by-screen pager, MORE-WORDS ( -- ). If you want a pager to use with
your own MWORDS word, you can use MORE ( xt -- ), e.g. as ' MWORDS MORE FOO
. (Note that I don't know how you will implement VOC-MWORDS, since as I
mentioned wordlists/modules don't really have names in zeptoforth.)

Travis

**Reply by ghost** [reply]
Yeah - will think about it in the near? future. Sometimes I ran out of time. :-(
My first approach would be to have a close look at word-info, and discard all the words that don't match a given wordlist.
Martin

**Comment by tabemann** [comment]
On Mon, Nov 4, 2024 at 9:59 AM MB ***@***.***> wrote:

> Yeah - will think about it in the near? future. Sometimes I ran out of
> time. :-(
> My first approach would be to have a close look at word-info, and discard
> all the words that don't match a given wordlist.
> Martin
>
WORDS-IN and MORE-WORDS-IN effectively do that -- they search the entire
flash dictionary and RAM dictionary of the main task for all words whose
wordlist field matches the specified wordlist/module except for those whose
names begin and end with * (except for * itself).

Travis
